### ISSUE TITLE
Implement convolve2d()

### ISSUE THREAD

**Issue by ClaudiaComito** [helmholtz-analytics/heat#920]
**Feature functionality**
https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.convolve2d.html#scipy.signal.convolve2d

**Additional context**
Signal processing / RFI mitigation for the SKA prototype dish, with @hrkloeck, @TobiasWinchen

**Comment by lolacaro** [comment]
I would take this over, okay?

**Comment by lolacaro** [comment]
The current implementation utilizes a locally written padding options which also enables symmetric or circular padding (not just by constant value). Should I remove this for now and just go with zero padding as for 1D convolutions, to be consistent? Or should I expand the 1D convolution to utilize the padding function available instead?

**Comment by JuanPedroGHM** [comment]
> The current implementation utilizes a locally written padding options which also enables symmetric or circular padding (not just by constant value). Should I remove this for now and just go with zero padding as for 1D convolutions, to be consistent? Or should I expand the 1D convolution to utilize the padding function available instead?

Expanding to 1D should be trivial, right? The padding function from torch already supports all the padding types: https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.pad.html

**Comment by lolacaro** [comment]
@JuanPedroGHM Yes, that can be done.
However, I have to check if my computation in 2D and 1D still holds because in this implementation, the padded array will not be balanced. I'll see if I can expand this.

**Comment by lolacaro** [comment]
The padding function implemented here is not trivial, and I think a discussion on how to proceed might be a good idea.
```python
def convgenpad(a, signal, pad, boundary, fillvalue):
    """
    Adds padding to local PyTorch tensors considering the distributed scheme of the overlying DNDarray.

    Parameters
    ----------
    a : DNDarray
        Overlying N-dimensional `DNDarray` signal
    signal : torch.Tensor
        Local Pytorch tensors to be padded
    pad: list
        list containing paddings per dimensions
    boundary: str{‘fill’, ‘wrap’, ‘symm’}, optional
        A flag indicating how to handle boundaries:
        'fill':
         pad input arrays with fillvalue. (default)
        'wrap':
         circular boundary conditions.
        'symm':
         symmetrical boundary conditions.
    fillvalue: scalar, optional
         Value to fill pad input arrays with. Default is 0.
    """
    dim = len(signal.shape) - 2
    dime = 2 * dim - 1
    dimz = 2 * dim - 2
    # check if more than one rank is involved
    if a.is_distributed() and a.split is not None:
        # set the padding of the first rank
        if a.comm.rank == 0:
            pad[dime - 2 * a.split] = 0
        # set the padding of the last rank
        elif a.comm.rank == a.comm.size - 1:
            pad[dimz - 2 * a.split] = 0
        else:
            pad[dime - 2 * a.split] = 0
            pad[dimz - 2 * a.split] = 0

    if boundary == "fill":
        signal = fc.pad(signal, pad, mode="constant", value=fillvalue)
    elif boundary == "wrap":
        signal = fc.pad(signal, pad, mode="circular")
    elif boundary == "symm":
        signal = fc.pad(signal, pad, mode="reflect")
    else:
        raise ValueError("Only {'fill', 'wrap', 'symm'} are allowed for boundary")

    return signal

```


**How the padding is implemented:**
- You can be certain here that `a` has two dimensions and `signal` has 4 dimensions, where the last two match `a`
- `torch.nn.functional.pad` takes a pad list like `[pad_last_0, pad_last_1, pad_second2last_0, pad_second2last_1]`, where 0 and 1 give left and right or up and down padsizes
- The function `convgenpad` in `signal.py` checks this padding tensor against the split dimension and for example, if it is split on axis 0 and the rank is 0, `pad_second2last_1` is set to 0 to ensure that locally the tensor is only padded on the correct side of the local tensor  (opposite if the rank is the maximum rank, inbetween both are set to 0)
- Then apply padding locally using either `constant`, `circular`or `reflect` boundary condition

**What is the problem?**
- `circular` boundary conditions on rank 0 need to communicate with the last rank to get the correct padding information, which is not given in this implementation
- `reflect` can work but only if the added padded size is smaller than what is available locally (including halo), this is not checked. However, this practically can only occur if the kernel is distributed, because:
1. if the kernel is not distributed, the kernel is always smaller than the local signal chunk and the pad size never exceeds `kernel_size -1`
2. if the kernel is distributed, this can happen (? please, correct me if wrong), because it is only ensured that the local kernel chunks are smaller than the local signal chunk, but the padding is based on the total kernel length

_Interesting side note:_
- Boundary conditions `constant` and `replicate` will work this way, but `replicate` is not yet an option.

**What to do now?**
I see three ways forward
1. Only enable constant padding and use the heat pad function
2. Use `convgenpad`but restrict it to boundary conditions `constant` and `replicate`, and maybe enable `reflect` only if the local size is larger than the padded area. And if you like `circular` or  a general padding function with all these options open an issue on the topic.
3. I go on a side quest and start implementing a general padding function before coming back to convolutions.

**Comment by lolacaro** [comment]
I correct myself.
`reflect` should always work, because we ensure that the kernel always fits into the local distributed signal chunk. Hence even if we pad with `kernel_length -1` there is just enough to reflect on.
